### PR TITLE
Phase2 workflow revamp

### PR DESCRIPTION
**Pull request by kpedro88**
This PR reorganizes the Phase2 workflows as discussed in the Upgrade Studies meeting, in preparation for HGCal "cheated" reco.

The new naming scheme is described in [Configuration/Geometry/README](https://github.com/kpedro88/cmssw/blob/Phase2-WF12/Configuration/Geometry/README.md) as requested by @ianna. There is also a new [Configuration/Geometry/test](https://github.com/kpedro88/cmssw/tree/Phase2-WF12/Configuration/Geometry/test) with a script to generate 2023 geometries automatically. (For the Phase2 Eras, the only difference is caused by the calorimeter content, so I named them `Phase2C1` and `Phase2C2`.)

The upgrade workflow matrix numbers have been reorganized. This is reflected in all the necessary configuration files.

Other items of note:
- `run2_common` is now chained into the Phase2 Eras
- `hbheUpgradeReco` and `hfUpgradeReco` are no longer used as tags; the reco modules themselves are now replaced using `toReplaceWith`, so the usual `hbhereco` and `hfreco` tags for the RecHit collections will continue to work in all downstream reco
- phase2 validation and DQM sequences are now defined in the autoValidation and autoDQM configs (@davidlange6, let me know if this is what you had in mind)
- upgrade workflow components are reorganized slightly (@boudoul, let me know if you have any qualms - I just wanted to have one dictionary for everything)

All the workflows should still run. I demoted the HGCal workflows back to local reco until the cheated reco is fully ready (@lgray is working on the necessary downstream reco changes).

Don't be intimidated by the apparent size of this PR; ~98% of the changes are in Python...

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by mmusich** [comment]
+1 
for changes in ` Calibration/IsolatedParticles/plugins/HcalRaddamMuon.cc `

**Comment by kpedro88** [comment]
@mdhildreth if you could also look at this once (mostly from the geometry side), that would be appreciated

**Comment by slava77** [comment]
with simple runTheMatrix.py I see 
- 10621++: D1 PU
- 10821++: D2
- 11021++: D2 PU
- 11221++: D3
Please add D1 (needed for simpler/fast validation), which are 10421++ in the "-w upgrade".

The "-l limited" suggests you wanted to have 10424.0 in the non-upgrade setup.

**Comment by kpedro88** [comment]
@cmsbuild please test

@slava77 apparently I forgot to commit the update to relval_2023.py, thanks for the catch. Fixed now.

File: Configuration/Geometry/test/dict2023Geometry.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+    }
+}
+
+allDicts = [ commonDict, trackerDict, caloDict, muonDict, forwardDict ]
+
+detectorVersionDict = {
+    ("O1","T1","C1","M1","F1") : "D1",
+    ("O1","T2","C1","M1","F1") : "D2",
+    ("O1","T1","C2","M1","F1") : "D3",
+}
```
Comment (slava77) [inline Configuration/Geometry/test/dict2023Geometry.py]: newline

**Comment by kpedro88** [comment]
@slava77 let me know when you are finished with code review so I can address it all at once

**Comment by slava77** [comment]
Nothing from visual inspection.
I will check configs and results by tomorrow.

**Comment by slava77** [comment]
Just an observation of a change in line with discussion preceding this PR.

The old short matrix 11024 "LReco" workflow (the one with HGCAL) was coupled with Phase2LReco era and cust_2023LReco (= customiseFlatTK) customization and Extended2023LReco (flat tracker) geometry.
The new short matrix  workflow replacint 11024 is 11224 "LReco" workflow (the one with HGCAL) is coupled with Phase2C2 era and cust_2023tilted customization and Extended2023D3 geometry.

**Comment by kpedro88** [comment]
@slava77 yes, that's correct. Hopefully the customization function will be going away soon.

File: Validation/Configuration/python/autoValidation.py [hunk 2]
```diff
@@ -6,3 +6,4 @@
                    'standardValidation' : ['prevalidation','validation','validationHarvesting']
                  }
 
+autoValidation['phase2Validation'] = autoValidation['baseValidation']+autoValidation['trackingOnlyValidation']+autoValidation['muonOnlyValidation']
```
Comment (slava77) [inline Validation/Configuration/python/autoValidation.py]: apparently this syntax doesn't work. cmsDriver output with this auto ends up having only the first element baseValidation, while everything else is lost.

File: DQMOffline/Configuration/python/autoDQM.py [hunk 3]
```diff
@@ -49,3 +49,4 @@
                           'dqmHarvesting']
             }
 
+autoDQM['phase2'] = autoDQM['trackingOnlyDQM']+autoDQM['muon']
```
Comment (slava77) [inline DQMOffline/Configuration/python/autoDQM.py]: this didn't work either: muon dqm gets lost

**Comment by slava77** [comment]
I looked at configs in 10624 (baseline) vs 10424 (this PR) and see
- changes related to run2_common < -- expected
- loss of dqm and validation sequences <-- I'm guessing, not expected.

**Comment by kpedro88** [comment]
@slava77 I will try to figure out the problem with DQM/validation this afternoon

**Comment by kpedro88** [comment]
@cmsbuild please test

@slava77 figured out the problem, just needed to compose the sequences differently

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/15393	 7236319
- changes are in line with the description
- jenkins tests pass and show no changes in pre-phase2 workflows
- local tests of equivalent 2023 workflows show changes apparently driven by the shift to run2_common (this triggers changes as early as in SIM. So, the geant random sequences are not the same and comparisons with baseline show a lot of changes). Config inspection suggests the changes are as intended.

**Comment by davidlange6** [comment]
Hi all - as already discussed in the ORP, please make any further comments on this PR today. Thanks

**Pull request merged**